### ISSUE TITLE
CT: sprite instance not found for selected bunny

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/natural-selection#138]
After fixing #136, there's a new PhET-iO issue:

```
natural-selection : phet-io-state-fuzz : unbuilt
https://bayes.colorado.edu/continuous-testing/ct-snapshots/1595390064648/phet-io-wrappers/state/?sim=natural-selection&phetioDebug&fuzz&wrapperContinuousTest=%7B%22test%22%3A%5B%22natural-selection%22%2C%22phet-io-state-fuzz%22%2C%22unbuilt%22%5D%2C%22snapshotName%22%3A%22snapshot-1595390064648%22%2C%22timestamp%22%3A1595405800237%7D
Uncaught Error: Assertion failed: sprite instance not found for selected bunny
Error: Assertion failed: sprite instance not found for selected bunny
at window.assertions.assertFunction (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1595390064648/assert/js/assert.js:22:13)
at NaturalSelectionSprites.setSelectedBunny (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1595390064648/natural-selection/js/common/view/environment/NaturalSelectionSprites.js:346:17)
at https://bayes.colorado.edu/continuous-testing/ct-snapshots/1595390064648/natural-selection/js/common/view/environment/NaturalSelectionSprites.js:179:63
at TinyEmitter.emit (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1595390064648/axon/js/TinyEmitter.js:69:53)
at SelectedBunnyProperty._notifyListeners (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1595390064648/axon/js/Property.js:273:25)
at PhaseCallback.listener (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1595390064648/axon/js/Property.js:323:47)
at PropertyStateHandler.attemptToApplyPhases (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1595390064648/axon/js/PropertyStateHandler.js:264:41)
at PropertyStateHandler.undeferAndNotifyProperties (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1595390064648/axon/js/PropertyStateHandler.js:198:12)
at https://bayes.colorado.edu/continuous-testing/ct-snapshots/1595390064648/axon/js/PropertyStateHandler.js:72:12
at TinyEmitter.emit (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1595390064648/axon/js/TinyEmitter.js:69:53)
id: Bayes Chrome
Snapshot from 7/21/2020, 9:54:24 PM
```

**Comment by pixelzoom** [comment]
This is occurring in the NaturalSelectionSprites.js, in the listener for `bunnyCollection.selectedBunnyProperty`.    "sprite instance not found for selected bunny" would seem to indicate that `selectedBunnyProperty` is being set before `bunnyCollection.bunnyCreatedEmitter` is firing (which creates the sprite associated with the bunny).     But I have not been able to confirm, and I have not been able to reproduce locally with phet-io-wrappers/state/?sim=natural-selection&phetioDebug&fuzz.

**Comment by pixelzoom** [comment]
Steps to reproduce manually:

1. Add a `debugger` statement to assert.js, just before `throw new Error`.
2. Run the unbuilt sim in the State Wrapper with `?sim=natural-selection&phetioDebug`
3. Go to the Intro screen.
4. Select the lone bunny.  This will trigger the assertion failure:

```
?sim=natural-selection&phetioDebug:115 Uncaught Error: Assertion failed: sprite instance not found for selected bunny
    at window.assertions.assertFunction (assert.js:23)
    at NaturalSelectionSprites.setSelectedBunny (NaturalSelectionSprites.js:346)
    at NaturalSelectionSprites.js:179
    at TinyEmitter.emit (TinyEmitter.js:69)
    at SelectedBunnyProperty._notifyListeners (Property.js:273)
    at PhaseCallback.listener (Property.js:323)
    at PropertyStateHandler.attemptToApplyPhases (PropertyStateHandler.js:264)
    at PropertyStateHandler.undeferAndNotifyProperties (PropertyStateHandler.js:198)
    at PropertyStateHandler.js:72
    at TinyEmitter.emit (TinyEmitter.js:69)
```

**Comment by pixelzoom** [comment]
Inspecting variables at the site of the assertion failure in NaturalSelectionSprites.js ...

`this.bunnyCollection.selectedBunnyProperty` has the expect value -- the bunny with tandem "bunny_0".

`this.spriteInstances` contains 6 elements, all of type ShrubSpriteInstance. So no BunnySpriteInstance has been created yet.  This means that `selectedBunnyProperty` has been restored before the `bunnyCollection.bunnyCreatedEmitter` fired to create the sprite for "bunny_0".

**Comment by pixelzoom** [comment]
@zepumph and I paired on this.  

First, we discovered that `bunnyCollection.bunnyCreatedEmitter` is no longer needed. Adding a listener to `bunnyCollection.liveBunnies.addItemAddedListener` does the job.  So that change was made in the above commit, and I think it's a nice improvement.

Next, we investigated the state problem.

As I hypothesized above, the problem is that `selectedBunnyProperty` is restored before `bunnyCollection.liveBunnies` listeners are fired.   And there doesn't seem to be a way to specify an order dependency in this case.  

So in the above commit, we added code that is specific to `phet.joist.sim.isSettingPhetioStateProperty`.  It cause `setSelectedBunny` to bail if the bunny's BunnySpriteInstance  doesn't exist yet, and it makes `createBunnySpriteInstance` responsible for calling `setSelectedBunny` in that case.  I'm not wild about this solution, because it introduces code paths that will be exercise only by the State Wrapper. 

The scenario of a PhetioGroup with an associated selectedProperty seems like something we're going to encounter often. So perhaps there should be a better solution. @zepumph will create a general issue to discuss with the iO team.

I'll leave this open until CT expresses its appreciation for this fix.
